Add a filtering API to VoiceCAllHandler

dcaliste opened the pull request:
Following @abranson comment from #14 , I'm proposing an API extension allowing to filter out incoming calls. This API is adding a new method to `AbstractVoiceCallHandler` interface. Besides the existing `answer()` for an incoming call, I propose to add `filter(action)`. Calling the later, allow the providers (telepathy or oFono) to deal with the call (by hanging up or letting it ring) and change its status. It is thus adding two new status: REJECTED and IGNORED. Introducing these two new status allow not to change much from `voicecall-ui`. Indeed, in case of filtering, an incoming status will be changed for rejected or ignored, so the ui won't show up and it will not call `playRingTone()`, which is called only for the incoming status.

This is a clean solution that avoid to call playRingTone() and immediately after to call silenceRingTone() or stop it because the call has been ignored or hang up. It still has the following drawback compared to the initial #14 proposition:
- the ending call dialog is still shown. But this can be now tuned in the UI easily since we have the new status that tells that it's not an incoming call. Some slight adjustements should be done in voicecall-ui not to show the ending call dialog for these new status.
- the calls are still seen as missed call or plain call in the call history. Indeed, commhistory-daemon is listening via Telepathy, which cannot flag calls (as far as I know). I'm wondering if it could be a good idea to remove the telepathy code from commhistory-daemon and use voicecallmanager instead. It would simplify the code in my opinion and avoid some duplicates between commhistory-daemon and voicecallmanager.

Regarding the API itself, it is proposing the same filtering capabilities as the ones exposed by oFono and the ones from Android call screening response API (as pointed out by @abranson). Compared to Android, I didn't add the possibility to mess up with the call log and notifications. I think that the first is not a good idea (I don't want to have calls filtered out without having the possibility to supervise what happened), and I think that the second belongs to voicecall-ui, based on the new status. About the name, I remember that @pvuorela pointed out that `filter` might not be a good name, since one can confuse the action with filtering / sorting results. Android is using "screening" as a term. What do you think ? Should the new method be called "screenIncomingCall()" or something like that ?

I'm keeping the second commit implementing a filtering plugin based on this API, to show how it can be used. Depending on the discussion, I will close #14 if this PR is found to be a better approach.

dcaliste committed:
[voicecall] Add a filtering plugin for incoming calls.

The plugin is based on a list of blacklisted
numbers in DConf. When a call is coming with
a matching lineId, the call is hang-up or
ignored automatically.

dcaliste committed:
[voicall] Add a filtering action for incoming calls.

When an incoming call arrives, besides existing
answer() and hangup() responses, this commit
adds a new one: filter(action). The action
filter() is doing, can be to block the call
(hangup automatically) or to ignore the call
(call is not hangup but no feedback is given).

This introduce two new states, REJECTED and
IGNORED.

attah commented:
Not that you should care very much about what i think: but i definitelyprefer filtering over screening. The latter, as a synonym "filtering", is usually not well understood for non-native speakers. Screening tends to be understood as a synonym to investigate, survey etc. How about something based around "blocking"?

abranson commented:
This looks good. The distinction between rejecting or just muting calls is important. I wonder if muting a call and letting it go to voicemail counts as a positive connection and encourages spammers, but I don't think we have a way to block that for the current call. I also wonder if that's still a thing, or they're just spamming hundreds of numbers regardless.

I sent a message to the Phonehook developer about this. They've been dealing with this for years and if anyone would know exactly what this API needs, it's them.

dcaliste commented:
Thank you @attah and @abranson for your comment. About the naming, I'm afraid "blocking" is too restrictive as a name, since one can also simply ignore the call.

Thank you @abranson for contacting Phonehook developers. I several times looked at their code to see how they are doing things. I cannot say that I read it all though : / They are using an SQlite backend for filtering storage, they can filter on contact or on number, as far as I understand. I'm wondering if the usage of an SQlite storage is required, if it ends up with so many numbers that going through all of them becomes a bootleneck. The disavantage I see to it is that, there is no automatic signaling when the database is changed. So a frontend adding a filtered number for instance as no simple mechanism to notify a voicecallmanager plugin of the change. This is not an issue in Phonehook case, but it could be one if we separate UI and filtering. That's why I chose to use a DConf entry. But I'm wondering if it can become too slow... The blocking in Phonehook is then done via DBus, calling hangup on the voicecallmanager object over the bus. That's why, in this PR, I've also added the "filter(action)" API to DBus (and not only internally for sync plugin usage). So like that, Phonehook can be updated to use this new API, if they desire so. Last thing about Phonehook, I'm still wondering why they are patching Lipstick, but that's because I've not read it carefully enough yet !

pvuorela commented on src/dbus/voicecallhandlerdbusadapter.h:
```diff
@@ -81,6 +81,7 @@ class VoiceCallHandlerDBusAdapter : public QDBusAbstractAdaptor
 public Q_SLOTS:
     bool answer();
     bool hangup();
+    bool filter(AbstractVoiceCallHandler::VoiceCallFilterAction action);
```
Intentionally added to d-bus interface but missing from the declarative side that does the d-bus calls?

dcaliste replied:
Yes, the addition to the voicecall-manager daemon D-Bus API is to be able to use the filtering capability from an external program like phone-hook. Currently, I created two ways to filter a call:
- via a plugin, then D-Bus is not involved,
- via a D-Bus call on "/calls/xxx" object path of the `voicecall-manager` daemon.

If one wants to do it via QML, like in the voice-call-ui for instance, then, the code will have to be added in the declarative plugin, you're right.

pvuorela commented on src/dbus/voicecallhandlerdbusadapter.cpp:
```diff
@@ -215,6 +215,19 @@ bool VoiceCallHandlerDBusAdapter::hangup()
 }
 
+/*!
+  Initiates filtering this voice call, if its' an incoming call.
```
Repeats on earlier, but these should be all "it's", right?

dcaliste replied:
Fixed.

pvuorela reviewed (commented):
Going through again the existing code in this voicecall projects there are details that feel suspicious, like registered *handlers* reported by voiceCallAdded. Then the actions done on the handler, not a phone call, which itself doesn't have any specific handle to use for referring.

The changes here slightly continuing on that, filtering the handler instead of the call, though granted that anything better would need further changes. Then again by such I'm not sure how much the handler should be needing information on filtered calls instead of just being asked to hang up etc. At the moment there doesn't seem to be much other actions.

Another thing I'm pondering is whether this is racy. Now the order of d-bus signalling for the incoming call to voicecall-ui can first get the incoming call notification and slightly after that message for "nevermind, we ignore this"(?)

With these, still thinking would it be better or worse handling this in the ui app level.

> the calls are still seen as missed call or plain call in the call history. Indeed, commhistory-daemon is listening via Telepathy, which cannot flag calls (as far as I know). I'm wondering if it could be a good idea to remove the telepathy code from commhistory-daemon and use voicecallmanager instead. It would simplify the code in my opinion and avoid some duplicates between commhistory-daemon and voicecallmanager.

That seems indeed problematic. There might be some chance to adjust the status of the commhistory items but generally would indeed sound better if there was some middle layer feeding the call items and commhistoryd not listening them itself.

dcaliste commented:
I apologise for taking such a long time to answer your review. As you've seen, I've tested various things with libcommhistory and here at voicecall, making related but independant commits on DBus related parts or call handler workflow. Now, I'm coming back to this PR.

> Going through again the existing code in this voicecall projects there are details that feel suspicious, like registered handlers reported by voiceCallAdded. Then the actions done on the handler, not a phone call, which itself doesn't have any specific handle to use for referring.

I'm sorry, I don't get what you mean here :/ The implementations of the voicecallhandler interface (in ofono or telepathy providers) are phone calls and have ways to refer to the actual provider call (either via telepathy or directly via QOfono).

> Then again by such I'm not sure how much the handler should be needing information on filtered calls instead of just being asked to hang up etc. At the moment there doesn't seem to be much other actions.

The call handler have several actions like `hangup()`, `hold()`, `answer()`, `split()` or `merge()`. I'm introducing a new one `filter()` that takes a parameter with three possible actions : ring (normal case), reject (= hangup) or ignore (= let the call continue but don't advertise it to the device user). Reject must be different from hangup() since one will later need to log the call differently. I agree this is not visible in PR at the moment because I'm still wondering how to convey this information to the comm history. But at least, it's changing the status of the call handler.

> Another thing I'm pondering is whether this is racy. 

That's correct. I've reworked the PR to avoid it. To do so, I've updated the providers not to change the status to STATUS_INCOMING immediately for a newly created voicecallhandler. It relies on calling `filter()`on it later to set the status to INCOMING, REJECTED or IGNORED. I've added a `setCallFiltering()` method in the manager also. When not set, the manager is always calling `filter(CONTINUE)` on any new call handler, keeping the old behaviour. But when setCallFilter() is set, the call is added to the manager and advertised over D-Bus and so on, but it stays in the NULL state as long as no one called `filter()` on it.

It's actually a similar behaviour than the one used by Slava in its filtering implementation in oFono: a call is not in an active state is there is a possible filtering, waiting for the filtering decision.

pvuorela commented on plugins/providers/telepathy/src/streamchannelhandler.cpp:
```diff
@@ -774,6 +772,12 @@ void StreamChannelHandler::setStatus(VoiceCallStatus newStatus)
     if (newStatus == d->status)
         return;
 
+    if (newStatus == STATUS_DISCONNECTED &&
```
By convention operators to beginning of lines.

dcaliste replied:
Sure, corrected.

pvuorela commented on plugins/filter/src/filterlist.cpp:
```diff
@@ -0,0 +1,67 @@
+
+bool FilterList::match(const QString &number)
+{
+    return list().contains(number);
```
It's a start, but ideally the phone numbers should be matched a bit more smart to make sure numbers with and without the plus prefixed country code get a match.

Maybe qtcontacts-sqlite / qtcontacts-extensions_impl.h might help here, libcommhistory also having some wrapper. Don't remember immediately how this should all go.

dcaliste replied:
Indeed, it's just a placeholder for the moment, working more on the changes required in the stack than the implementation of the blocking itself. I'm sure I'll need something like a regexp anyway since French spammers are using ranges of phone numbers.

I was thinking about libphonenumber to do the actual job of matching numbers. But I didn't look at all at their API for the moment.

pvuorela replied:
The regexp is probably too techy to expose on the UI so if wanted that could be maybe a different config key or even a separate plugin.

attah replied:
I was going to offer some slightly too generic guesses for what is needed (exact and prefix matching only - after normalization). Heck, even exact only might be fine just to get it out the door.
Then i started looking at the libphonenumber... 
I found found some red herrings (e.g. matching number inside text).
Then i found something decently promising (RegexBasedMatcher::Match with allow_prefix_match), only to find that it's not part of the installed headers.
Nor are the building blocks to remake it despite it being trivial. :(

Now i just want to go cry in a corner and feel ashamed to prefer C++ over C.
What a horrible mess they have made. So much noise for basically no functionality beyond the parsing.

...so looping back around. Exact match is done with:
```c++
PhoneNumberUtil* util = PhoneNumberUtil::GetInstance();
util->IsNumberMatchWithTwoStrings(number1, number2)==PhoneNumberUtil::EXACT_MATCH
```
It does not have a lot of wiggle-room/help in the parsing from what i can understand - so what is the point?
Are we just opening up to inability to compare (and thus block) number formats that the current version of libphonenumber considers valid?
(see: https://forum.sailfishos.org/t/cant-send-sms-when-number-too-long/18544)

dcaliste replied:
> The regexp is probably too techy to expose on the UI 

Yes sure. I was not clear in my reply. I don't mean to expose regexp in the UI. That's the beauty of discoupling engine and UI. The filtering plugin can handle regexp while the UI is exposing something simple like a context menu in phone log saying "block this number" for instance. Or one can even provide a dedicated "advanced" application that can expose simple matching like suggested by @attah with prefix matching...

The current plugin is based on DConf values. It's easy to implement something that can handle:
```
ignored_numbers = ["+33123456789", "regexp:^{\+33,0}1234567[0-9][0-9]"]
whitelist_numbers = ["+33123456700", "0123456700"]
```
A simple UI integration like the one mentioned before can add or remove numbers like "+33123456789", while an advance app or a techy user can add more complex patterns.

@attah, thanks for your look into libphonenumber. It doesn't sound promising for our application. Seems more oriented for parsing or validating numbers. Doesn't matter, using Qt regexp internally can be largely sufficient after all.

pvuorela commented on rpm/voicecall-qt5.spec:
```diff
@@ -123,3 +131,7 @@ fi
 %{_libdir}/voicecall/plugins/libvoicecall-ofono-plugin.so
 
+%files plugin-voicecall-filter
+%defattr(-,root,root,-)
```
Exists already on other packages, but no need to have these defattrs anymore, it's the default.

dcaliste replied:
Sure, line removed. I'll try to remember it for next time.

pvuorela reviewed (commented):
> I'm sorry, I don't get what you mean here :/ The implementations of the voicecallhandler interface (in ofono or telepathy providers) are phone calls and have ways to refer to the actual provider call (either via telepathy or directly via QOfono).

Not sure either myself anymore :) Generally separate handler for each call should identify the calls good enough. Maybe was thinking conference call, but I'm not sure is that too much involved in the filtering here.

But generally this seems to mostly do the trick and not being too big of a change, so I'm warming up to just have this for start, and then maybe even do some simple UI on the voicecall.

Still thinking, though, if ideally the UI app should be in control of filtering, and also on what gets added to the call log (meaning commhistoryd not being responsible for listening to telepathy). Or if not UI, could be better even if this component did the call log, like maybe marking a filtered call as filtered. Commhistoryd can't know at the time it makes the log item if it's being blocked.

dcaliste commented:
> Still thinking, though, if ideally the UI app should be in control of filtering, and also on what gets added to the call log (meaning commhistoryd not being responsible for listening to telepathy). Or if not UI, could be better even if this component did the call log, like maybe marking a filtered call as filtered. 

Yes, that's exactly my reasoning. I've started some early attempts here and there. Which provides some feedback:
- dropping telepathy observers from commhistoryd in favour of a listener on voicecall, would loose the genericity of it. Some says that Telepathy is dead so maybe there will be no gain in keeping it. But in principle it's nice that one can log SIP and GSM calls with the same code. In practice Telepathy adds quite a large code overhead and there is a large code duplication between voicecall/providers/telepathy and telepathy observers in commhistoryd. So I'm mitigated on keeping it in commhistoryd or not.
- the simplest stuff is to implement logging as a plugin in voicecall (like the mce or ngfd ones), using libcommhistory to write to the database. It represent only 200 lines of code. That's nice, but a small mistake and the telephony crashes. Without speaking about the fact that it's synchronous with the call handling, so adding possible delays where we absolutely don't want any. So for the moment, I didn't fully debug my first attempt and didn't publish it yet.
- the other possibility is to move the logging code to commhistoryd, to replace part of the telepathy one. It can then listen to the D-Bus interface of voicecall. I think it's the best option, but I didn't try it yet for time reasons. And also because the D-Bus API from voicecall is not exposed, but only used internally for the QML plugin. So it requires some changes there also.

pvuorela commented:
> Some says that Telepathy is dead so maybe there will be no gain in keeping it.

If we'd implement these from scratch, I would just skip telepathy. But now that it's there let's not indeed go removing it as side-effect of work like this.

> That's nice, but a small mistake and the telephony crashes.

Well, then again the small mistake can be now in commhistoryd too. The services should get automatically restarted if they crash.

> Without speaking about the fact that it's synchronous with the call handling, so adding possible delays where we absolutely don't want any.

Not sure though can that block for a long time. In UI everything should be done in 16ms but voicecall being a background service it shouldn't matter that much if something takes a 100ms.

dcaliste commented:
I've rebased this PR. I've also updated it to use sailfishos/libcommhistory#10, so it can store in the event log the status of filtered out incoming calls.

I'm going to test it more extensively in the coming days.

pvuorela commented on plugins/commhistory/src/commhistoryplugin.cpp:
```diff
@@ -65,6 +65,11 @@ class CommHistoryPlugin::Private
                                                 : CommHistory::Event::Outbound);
+        if (handler.status() == AbstractVoiceCallHandler::STATUS_IGNORED) {
+            event.setIncomingStatus(CommHistory::Event::Ignored);
+        } else if (handler.status() == AbstractVoiceCallHandler::STATUS_REJECTED) {
```
Tried adding a phone number to blocked ones and made a call. 

Seems like this is reached with null status, then there's a onVoiceCallStatusChanged() with rejected but that doesn't now update the incoming status.

dcaliste replied:
Thanks for testing, I still need to do it myself. I'm a bit busy in the coming days, but should find time to do it properly early next week. I've just added these lines yesterday, to show how things are coming together and PRs are related. But I didn't have time yet to check that everything is properly working.

My early thoughts were that after the commit 2b1afb6dc5eedc91d59a4588b8ed7fdbeb290754, the first status should be INCOMING, REJECTED or IGNORED. But you're right, I forgot that first, status is NULL awaiting for filter decision. I remember it now :/ So I need to implement this a bit differently.

dcaliste commented:
I've updated the PR to use uniformally "rejected" everywhere, while before, the filter action was called "blocked" and the status was called "rejected".

I've quickly tested it with the associated PR in libcommhistory, ignored calls are properly saved in the commhistory as an event property (`40069|incomingStatus|ignored` for instance), and the phone is not ringing (but announce for an ongoing call, which is right : the modem is in use). Normal calls are still ringing and when not answered are labbeled as missed and a notification is created.

Some strange behaviour : calling from an ignored number is adding an incoming call in the log (which is fine for the moment since the call history list UI has not been updated to take into account the new incoming status). Then, calling from the same number (after its removal from the ignored list) but not answering is creating a missed call notification (which is fine also), but the call history is clustering the two calls with a miss call symbol and a little '2' figure, and put it in the highlight colour. I'm not sure it's an expecting behaviour since there is no two missed calls, only one. I don't know if it's related to this PR, I'll need to investigate.

pvuorela commented:
I can reproduce the strange behavior. I have also some local modifications for using an icon for blocked calls. Works otherwise fine, but first blocked call and then unblocked unanswered call has event count of 2. I think it should be showing just the latest unanswered unblocked call with count of 1. Guess something going wrong now in calculating the event count in libcommhistory side.

dcaliste commented:
Thanks for testing and pointing to libcommhistory. I'll give a look there to the event list models to see if I can find what's wrong with the case we're dealing with.

pvuorela reviewed (approved):
Works nice with exact plus prefixed phone number in reject config. There's still the detail of number matching, but if you don't mind I'm thinking we could start by just merging this now and handle the matching later.

Already did a bit of status to icon mapping on the UI side which I could merge then too. And continue on doing some UI for marking numbers as blocked. Thinking there we could now just do rejection, not confusing people with distinction between ignoring.

dcaliste commented:
Thank you for the apreciation. About your remarks, that's exactly my line of thinking also. I've added sailfishos/commhistory-daemon#17 not to notify on blocked calls (after all you just want not to be disturbed by them, having them in the call log is already enough to supervise what happened).

> and handle the matching later.

Yes, I think that separated PRs to improve the matching will be better now that the framework is present. I've got several ideas already (prefix matching, wild cards and whitelisting), but wanted to work on the infra-structure first.

> Thinking there we could now just do rejection, not confusing people

Exact, the UI should be simple enough. The separation in DConf can be used by advanced users with CLI, or even by people creating dedicated applications to tune the DConf values. I was thinking for instance to a use case where an application can change the values dynamically on a schedule for instance, blocking all numbers at night, except for the favourite contacts.
